src/semaphore.c

```diff
@@ -804,3 +889,110 @@ _dispatch_thread_semaphore_wait(_dispatch_thread_semaphore_t sema)
 #error "No supported semaphore type"
 #endif
 }
+
+#if USE_FUTEX_SEM
+#define DISPATCH_FUTEX_NUM_SPINS 100
+#define DISPATCH_FUTEX_VALUE_MAX INT_MAX
+#define DISPATCH_FUTEX_NWAITERS_SHIFT 32
+#define DISPATCH_FUTEX_VALUE_MASK ((1ull << DISPATCH_FUTEX_NWAITERS_SHIFT) - 1)
+
+static bool _dispatch_futex_wait_slow(dispatch_futex_t dfx,
+		const struct timespec* timeout);
+
+DISPATCH_INLINE
+static int
+_dispatch_futex_syscall(int *futex_addr, 
+		int op, int val, const struct timespec *timeout)
+{
+	return (int)syscall(SYS_futex, futex_addr, op | FUTEX_PRIVATE_FLAG, val, 
+						timeout, NULL, 0);
+}
+
+/* all memory orders below were added with strict seq_cst semantics
+ * assumed to be the strongest
+ */
+
+// Atomic-decrement-if-positive on the semaphore value, with acquire MO if
+// successful.
+DISPATCH_INLINE
+bool
+_dispatch_futex_trywait(dispatch_futex_t dfx)
+{
+	uint64_t orig;
+	while ((orig = dfx->dfx_data) & DISPATCH_FUTEX_VALUE_MASK) {
+		if (dispatch_atomic_cmpxchg2o(dfx, dfx_data, orig, orig - 1, acquire)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+// Increments semaphore value and reads wait count in a single atomic
+// operation, with release MO. If wait count is nonzero, issues a FUTEX_WAKE.
+DISPATCH_INLINE
+bool
+_dispatch_futex_signal(dispatch_futex_t dfx)
+{
+	uint64_t orig;
+	do {
+		orig = dfx->dfx_data;
+		if (slowpath((orig & DISPATCH_FUTEX_VALUE_MASK) ==
+					 DISPATCH_FUTEX_VALUE_MAX)) {
+			DISPATCH_CRASH("semaphore overflow");
+		}
+	} while (!dispatch_atomic_cmpxchgvw2o(dfx, dfx_data, orig, orig + 1, &orig, release));
+	if (slowpath(orig >> DISPATCH_FUTEX_NWAITERS_SHIFT)) {
+		int ret = _dispatch_futex_syscall(&dfx->dfx_futex, FUTEX_WAKE, 1, NULL);
+		DISPATCH_SEMAPHORE_VERIFY_SYSRET(ret);
+	}
+	return true;
+}
+
+// `timeout` is relative, and can be NULL for an infinite timeout--see futex(2).
+DISPATCH_INLINE
+bool
+_dispatch_futex_wait(dispatch_futex_t dfx, const struct timespec *timeout)
+{
+	if (fastpath(_dispatch_futex_trywait(dfx))) {
+		return true;
+	}
+	return _dispatch_futex_wait_slow(dfx, timeout);
+}
+
+DISPATCH_NOINLINE
+static bool
+_dispatch_futex_wait_slow(dispatch_futex_t dfx, const struct timespec *timeout)
+{
+	int spins = DISPATCH_FUTEX_NUM_SPINS;
+	// Spin for a short time (if there are no waiters).
+	while (spins-- && !dfx->dfx_data) {
+		dispatch_hardware_pause();
+	}
+	while (!_dispatch_futex_trywait(dfx)) {
+		dispatch_atomic_add2o(dfx, dfx_data,
+							  1ull << DISPATCH_FUTEX_NWAITERS_SHIFT, relaxed);
+		int ret = _dispatch_futex_syscall(&dfx->dfx_futex, FUTEX_WAIT, 0, timeout);
+		dispatch_atomic_sub2o(dfx, dfx_data,
+							  1ull << DISPATCH_FUTEX_NWAITERS_SHIFT, acquire);
```

Comment:
the indent rule is two tabs so that should be:

```C
		dispatch_atomic_sub2o(dfx, dfx_data,
				1ull << DISPATCH_FUTEX_NWAITERS_SHIFT, acquire);
```

and same above